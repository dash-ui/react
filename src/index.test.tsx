/* jest */
import * as React from 'react'
import {renderHook, cleanup} from '@testing-library/react-hooks'
import {render as renderComponent} from '@testing-library/react'
import styles from '@-ui/styles'
import {
  DashProvider,
  Inline,
  useDash,
  useGlobal,
  useThemes,
  useVariables,
  useStyle,
  useStyles,
} from './index'

afterEach(() => {
  document.getElementsByTagName('html')[0].innerHTML = ''
})

const opt = (props = {}) => ({
  wrapper: (other) => <DashProvider {...props} {...other} />,
})

const renderFragment = (children = null, props = {}, options = {}) =>
  renderComponent(children, {
    wrapper: ({children}) => <DashProvider {...props} children={children} />,
    ...options,
  }).asFragment()

declare module '@-ui/styles' {
  interface DefaultVars {
    color: {
      primary: string
      secondary?: string
    }
  }

  interface DefaultThemes {
    light: DefaultVars
    dark: DefaultVars
  }
}

describe('<DashProvider>', () => {
  afterEach(cleanup)

  it('provides the default styles() configuration', () => {
    const {result} = renderHook(() => useDash(), opt())
    expect(result.current).toBe(styles)
  })

  it('provides a custom styles() configuration', () => {
    const myStyles = styles.create()
    const {result} = renderHook(() => useDash(), opt({dash: myStyles}))
    expect(result.current).toBe(myStyles)
  })

  it('works without a provider', () => {
    const {result} = renderHook(() => useDash())
    expect(result.current).toBe(styles)
  })
})

describe('<Inline>', () => {
  afterEach(cleanup)

  it('writes css', () => {
    expect(
      renderFragment(
        <Inline
          css={`
            display: block;
          `}
        />,
        {styles}
      )
    ).toMatchSnapshot()
  })

  it('writes css w/ nonce', () => {
    const myStyles = styles.create({nonce: 'E8gagwlWEGlgwel'})
    expect(
      renderFragment(
        <Inline
          css={`
            display: block;
          `}
        />,
        {dash: myStyles}
      )
    ).toMatchSnapshot()
  })

  it('writes css object', () => {
    expect(
      renderFragment(<Inline css={{display: 'block'}} />, {styles})
    ).toMatchSnapshot()
  })

  it('writes css callback', () => {
    const myStyles = styles.create({variables: {color: {primary: '#000'}}})

    expect(
      renderFragment(<Inline css={({color}) => `color: ${color.primary};`} />, {
        dash: myStyles,
      })
    ).toMatchSnapshot()
  })

  it(`doesn't write falsy css callback`, () => {
    const myStyles = styles.create()
    myStyles.variables({color: {primary: '#000'}})

    expect(
      renderFragment(<Inline css={''} />, {
        dash: myStyles,
      })
    ).toMatchSnapshot()
  })
})

describe('useGlobal()', () => {
  it('sets global styles with a string value', async () => {
    const myStyles = styles.create()
    const {unmount, rerender} = renderHook(
      () => useGlobal(`:root { --blue: #09a; }`),
      opt({dash: myStyles})
    )

    rerender()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(1)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot(
      ':root'
    )
    unmount()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)
    await cleanup()
  })

  it('sets global styles with a function value', async () => {
    const myStyles = styles.create()
    myStyles.variables({color: {primary: '#000', secondary: '#fff'}})
    const {unmount, rerender} = renderHook(
      () => useGlobal(({color}) => `body { background: ${color.primary}; }`),
      opt({dash: myStyles})
    )

    rerender()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(2)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot(
      'variables'
    )
    expect(document.querySelectorAll(`style[data-dash]`)[1]).toMatchSnapshot(
      'global'
    )
    unmount()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(1)
    await cleanup()
  })

  it('handles falsy values', async () => {
    const myStyles = styles.create()
    renderHook(() => useGlobal(false), opt({dash: myStyles}))
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)

    renderHook(() => useGlobal(0), opt({dash: myStyles}))
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)

    renderHook(() => useGlobal(null), opt({dash: myStyles}))
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)

    renderHook(() => useGlobal(''), opt({dash: myStyles}))
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)
    await cleanup()
  })
})

describe('useVariables()', () => {
  afterEach(cleanup)

  it('adds variables then cleans up', async () => {
    const myStyles = styles.create()
    const {unmount, rerender} = renderHook(
      () =>
        useVariables({
          color: {primary: '#000', secondary: '#fff'},
        }),
      opt({dash: myStyles})
    )

    rerender()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(1)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot(
      ':root'
    )
    unmount()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)
    await cleanup()
  })
})

describe('useThemes()', () => {
  afterEach(cleanup)

  it('adds variables then cleans up', async () => {
    const {unmount, rerender} = renderHook(
      () =>
        useThemes({
          dark: {
            color: {primary: '#000', secondary: '#fff'},
          },
          light: {
            color: {primary: '#fff', secondary: '#000'},
          },
        }),
      opt({dash: styles.create()})
    )

    rerender()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(2)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot()
    expect(document.querySelectorAll(`style[data-dash]`)[1]).toMatchSnapshot()
    unmount()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(0)
    await cleanup()
  })
})

describe('useStyle()', () => {
  afterEach(cleanup)

  it('creates a style and class', async () => {
    const {result} = renderHook(
      () => useStyle(({color}) => `color: ${color.primary};`),
      opt({
        dash: styles.create({
          variables: {
            color: {
              primary: '#000',
            },
          },
        }),
      })
    )

    expect(result.current()).toMatchSnapshot()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(2)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot()
    expect(document.querySelectorAll(`style[data-dash]`)[1]).toMatchSnapshot()
    await cleanup()
  })
})

describe('useStyles()', () => {
  afterEach(cleanup)

  it('creates styles and class', async () => {
    const {result} = renderHook(
      () =>
        useStyles({
          primary: ({color}) => `color: ${color.primary};`,
          secondary: ({color}) => `color: ${color.secondary};`,
        }),
      opt({
        dash: styles.create({
          variables: {
            color: {
              primary: '#000',
              secondary: '#fff',
            },
          },
        }),
      })
    )

    expect(result.current('primary', 'secondary')).toMatchSnapshot()
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(2)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot()
    expect(document.querySelectorAll(`style[data-dash]`)[1]).toMatchSnapshot()
    await cleanup()
  })
})
