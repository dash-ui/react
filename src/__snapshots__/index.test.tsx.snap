// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Inline> doesn't write falsy css callback 1`] = `<DocumentFragment />`;

exports[`<Inline> writes css 1`] = `
<DocumentFragment>
  <style>
    display:block;
  </style>
</DocumentFragment>
`;

exports[`<Inline> writes css callback 1`] = `
<DocumentFragment>
  <style>
    color:var(--color-primary);
  </style>
</DocumentFragment>
`;

exports[`<Inline> writes css object 1`] = `
<DocumentFragment>
  <style>
    display:block;
  </style>
</DocumentFragment>
`;

exports[`<Inline> writes css w/ nonce 1`] = `
<DocumentFragment>
  <style
    nonce="E8gagwlWEGlgwel"
  >
    display:block;
  </style>
</DocumentFragment>
`;

exports[`useGlobal() sets global styles with a function value: global 1`] = `
<style
  data-dash="-ui"
>
  
  body{background:var(--color-primary);}
</style>
`;

exports[`useGlobal() sets global styles with a function value: variables 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--color-primary:#000;--color-secondary:#fff;}
</style>
`;

exports[`useGlobal() sets global styles with a string value: :root 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--blue:#09a;}
</style>
`;

exports[`useStyle() creates a style and class 1`] = `"-ui-7y43y9"`;

exports[`useStyle() creates a style and class 2`] = `
<style
  data-dash="-ui"
>
  
  :root{--color-primary:#000;}
</style>
`;

exports[`useStyle() creates a style and class 3`] = `
<style
  data-dash="-ui"
>
  
  .-ui-7y43y9{color:var(--color-primary);}
</style>
`;

exports[`useStyles() creates styles and class 1`] = `"-ui-ga4wf1"`;

exports[`useStyles() creates styles and class 2`] = `
<style
  data-dash="-ui"
>
  
  :root{--color-primary:#000;--color-secondary:#fff;}
</style>
`;

exports[`useStyles() creates styles and class 3`] = `
<style
  data-dash="-ui"
>
  
  .-ui-ga4wf1{color:var(--color-primary);color:var(--color-secondary);}
</style>
`;

exports[`useThemes() adds variables then cleans up 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-dark-theme{--color-primary:#000;--color-secondary:#fff;}
</style>
`;

exports[`useThemes() adds variables then cleans up 2`] = `
<style
  data-dash="-ui"
>
  
  .-ui-light-theme{--color-primary:#fff;--color-secondary:#000;}
</style>
`;

exports[`useVariables() adds variables then cleans up: :root 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--color-primary:#000;--color-secondary:#fff;}
</style>
`;
